### PR TITLE
LD-estimation for LDpred2 using bigSNPR

### LINKED ISSUE #108: LDpred2: Using precomputed ld and correlation matrix
**Is your feature request related to a problem? Please describe.**
So far I've tested the LDpred2.R script on real data using the infinitessimal mode, which seemed to work, but the auto mode did not as there was some problem with infinite values in the correlation matrix. To investigate such errors when these are not replicable using test data becomes challenging due to that calculation of the matrix takes several errors. Therfore the calculation of this matrix should be separated from the `ldpred2.R` script. This would
- Make debugging easier
- Simplify usage (less flags to keep track of for the current step)
- Improve speed as the matrix can be resused when running different modes of LDpred2. There is the issue of which SNPs to use when calculating the correlation matrix when using different sumstats, but that can be addressed later.

**Describe the solution you'd like**
All functionality that relates to calculating the correlation matrix is moved to a script of its own. The process would instead look like:
1) `createBackingFile.R file.bed file.rds`
2) `calculateSnpCorrelation.R file.rds snpcorr.rds`
3) `ldpred2.R file.rds snpcorr.rds sumtats.txt output.txt`

[This R-script](https://github.com/privefl/paper-infer/blob/main/code/example-with-provided-LD.R) provides an example how to use the ld provided by the creators of LDpred. This approach can be implemented in (2). If appropriate those ld files (from the [creators tutorial](https://privefl.github.io/bigsnpr/articles/LDpred2.html)) could maybe be provided in this repository as well. I guess there are cases when you can't run step 2.

**Additional context**
I've seen some issues on the LDpred2 github of people having similar ideas to speed up calculation of PGS using different sumstats. The challinge is that the format of the matrix used to calculate the correlation seem challening to handle. Appears you would have to create a matrix in the FBM format for all the SNPs you need and then subset this matrix and convert to SFBM for each set of SNPs for a particular sumstat file. But I don't know how big deal it is for normal usage if the overlap varies.

### PR DESCRIPTION
**Pull request by deepchocolate**
Fixes #108 

Changes proposed in this pull request:

This adds a new script `calculateLD.R` using the R-packaged bigSNPR to estimate LD.
- Takes a bigSNPR genotype (.rds) file as input
- Filter SNPs to use when calculating LD by a list of SNPs or a sumstat file.
- Optional random sampling of individuals to use in estimation.
- Outputs one table ("map") with SNP LD and a correlation matrix for each chromosome (.rds) to use as input in `LDpred2.R`.
- Renamed flag `--geno-file` to `--geno-file-rds` in `LDpred2.R`

So far I've only tested this using the simulated data provided by the LDpred2 tutorial. The ldsc regression gives a slightly different estimate than the tutorial; `Intercept = 0.9502722 H2 = 0.3002906` versus tutorial `## 0.9496371 0.2981213`. Maybe something with the filtering differs as in the tutorial the LD estimation is performed after using `snp_match`. I tried to avoid using that here as performs more tasks than are actually used for the LD estimation.

Outside this I took the liberty of removing some output in the LDpred2 script to make the output more readable...

## Before submitting

- [x] I've read and followed all steps in the [Making a pull request](https://github.com/comorment/containers/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [ ] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/comorment/containers/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [x] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

### CODE REVIEW

**Comment by deepchocolate** [comment]
Should probably have suggested updates to the tutorial docs aswell

**Comment by espenhgn** [comment]
Thanks @deepchocolate!
I'll try and run these codes now. In the meantime, can you update the CHANGELOG.md and /usecases/LDpred2/README.md files a bit with this addition?

File: usecases/LDpred2_example/run.sh [hunk 1]
```diff
@@ -11,20 +11,23 @@ export DIR_SIF=$DIR_BASE/singularity
...
 export fileInputSumStats=$DIR_BASE/usecases/LDpred2_tutorial/tutorial_data/public-data3-sumstats.txt
 export fileOutputSNPR=$DIR_TESTS/data/public-data3.rds
+export fileKeepSNPS=/REF/hapmap3/w_hm3.justrs
+export fileOut=$DIR_TESTS/output/public-data.score
 
 # Create shortcut environment variable for Rscript 
 export RSCRIPT="singularity exec -B $DIR_BASE:$DIR_BASE -B $DIR_REFERENCE:/REF $DIR_SIF/r.sif Rscript"
 
 # The different modes to run
 LDPRED_MODES="inf"
```
Comment (espenhgn) [inline usecases/LDpred2_example/run.sh:28]: could test both modes, i.e., `LDPRED_MODES="inf auto"?

**Review by espenhgn — CHANGES_REQUESTED** [review]
Hi @deepchocolate
I could run the `calculateLD.R` script in isolation (e.g., with the arguments `$RSCRIPT calculateLD.R --geno-file g1000_eur_chr21to22_hm3rnd1.rds --file-ld-blocks LD_with_blocks_chr@.rds --file-ld-map map.rds --chr2use 21 22`), but had some issues with the `run.sh` testing script that needs to be resolved. Make sure it will run in a clean git environment (run `git clean -fd .` in the `LDpred2_example` directory to remove untracked files). 
There was also some issue with unzipping `data/ld/ ldref_with_blocks.zip`. Doing it manually allowed the remaining tests to run.

We need some explanations on how to use the `calculateLD.R` script in combination with the `ldpred2.R` in the LDpred2/README.md file as well, and also make sure that the provided example(s) are up to date.

File: usecases/LDpred2/calculateLD.R [hunk 2]
```diff
@@ -0,0 +1,114 @@
+# Calculate LD
+library(bigsnpr, quietly = T)
+options(bigstatsr.check.parallel.blas = FALSE)
+options(default.nproc.blas = NULL)
+library(argparser, quietly = T)
+library(stringr)
+
+par <- arg_parser('Calculate linkage disequillibrium (LD) using bigSNPr')
+# Mandatory arguments
+par <- add_argument(par, "--geno-file", nargs=1, help="Input .rds (bigSNPR) file with genotypes")
```
[the remark this answers is not in the document]
Comment (deepchocolate) [reply usecases/LDpred2/calculateLD.R:10]: I don't see any problem with renaming it

File: usecases/LDpred2_example/run.sh [hunk 3]
```diff
@@ -34,37 +37,80 @@ if [ $? -eq 1 ]; then echo "$dump"; exit; fi
...
-## TEST: Bad ldpred mode
-dump=$( { $LDP --ldpred-mode infinite $fileOutputSNPR $fileSumstats $fileOut.inf; } 2>&1 )
-if [ $? -eq 0 ]; then exit; fi
+echo "### Testing ldpred2.R using externally provided LD"
+# Download data if necessary
+FILE_LDREF=$DIR_TESTS/data/ld/ldref_with_blocks.zip
+FILE_LDMAP=$DIR_TESTS/data/ld/map_hm3_plus.rds
+if [ ! -f $FILE_LDREF ]; then 
+ echo "Downloading LD reference"
+ wget -O $FILE_LDREF "https://figshare.com/ndownloader/files/36363087";
```
[the remark this answers is not in the document]
Comment (deepchocolate) [reply usecases/LDpred2_example/run.sh:50]: Ah, will fix the directory issue!

File: usecases/LDpred2_example/run.sh (hunk 1 above)
Comment (deepchocolate) [reply usecases/LDpred2_example/run.sh:28]: Correct, will fix!

File: usecases/LDpred2_example/run.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (deepchocolate) [reply usecases/LDpred2_example/run.sh:50]: I've added READMEs in the missing directories so they should exist now.

**Comment by deepchocolate** [comment]
@ofrei If I remember correctly you suggested moving the imputation to the `createBackingfile.R` script at the last meeting? I cannot attend the meeting today unfortunately, but I think I can wrap this PR up this week

File: usecases/LDpred2_example/run.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (espenhgn) [reply usecases/LDpred2_example/run.sh:50]: tests now run on my end if you agree and merge #122 into this branch.

File: usecases/LDpred2_example/run.sh [hunk 4]
```diff
@@ -34,37 +37,80 @@ if [ $? -eq 1 ]; then echo "$dump"; exit; fi
...
+
+echo "TEST error: Bad imputation mode"
+dump=$( { $LDP --ldpred-mode inf --geno-impute bad-mode; } 2>&1 )
+if [ $? -eq 0 ]; then echo "No error received"; exit; fi
 
 ## TEST: Complete runs of --ldpred-mode given by $LDPRED_MODES
 for MODE in $LDPRED_MODES; do
  echo "Testing mode $MODE"
- dump=$( { $LDP --ldpred-mode $MODE $fileOutputSNPR $fileSumstats $fileOut.inf; } 2>&1 )
+ dump=$( { $LDP --ldpred-mode $MODE; } 2>&1 )
```
[the remark this answers is not in the document]
Comment (espenhgn) [reply usecases/LDpred2_example/run.sh:78]: resolving, think this is fixed.

**Review by espenhgn — CHANGES_REQUESTED** [review]
Thanks @deepchocolate,
I'm fine with this if you can first merge #122 into this branch (didn't want to push directly to your working branch).

As for where to put the imputation that you mentioned, we can have things as is here for now. We can create another [issue](https://github.com/comorment/containers/issues) for it.

**Review by espenhgn — APPROVED** [review]

**Review by ofrei — CHANGES_REQUESTED** [review]
@deepchocolate great work!!

``--sumstats / --sumstats-sep`` feature seem to be somewhat redundant as use can achieve the same effect using ``--file-keep-snps``, and a few simple bash commands, e.g. `` cut -d' ' -f2- `` to extract a specific column from sumstats file, followed by ``comm -12 <(sort file1) <(sort file2)`` to get intersection of lines from two files. I suggest to remove ``--sumstats`` and ``--sumstats-sep``  argument. But I'm also fine to keep both - in this case clarify in the documentation that ``--sumstats`` is used only to filter out snps, that it's optional, and that if it's provided in addition to ``--file-keep-snps`` then it's the intersection that will be used. Also consider renaming ``--file-keep-snps`` to ``--extract``  (a naming that comes from plink https://www.cog-genomics.org/plink/1.9/filter and I think most analysts in the field will find it convenient if software follows conventions set by plink). But that's totally optional, if you prefer current name ``--file-keep-snps`` is also fine!

https://privefl.github.io/bigsnpr/reference/snp_cor.html has thr_r2 argument, I think we should allow user to specify it (via additional argument to calculateLD.R). We can also set a default of 0.01 which saves a lot of disk space (and time to load the LD matrix, and most likely time used in LDpred2 computation if it benefits from sparse LD matrix). 


> Genetic distance is used by snp_cor, argument infos.pos. First, for some reason the documentation states that this is supposed to be physical position basepairs (thus integers). However, the tutorial uses genetic distance in centimorgans.

I think it's fine - both ways of restricting the "band" of the sparse LD matrix are valid.  There are 3 ways of defining the distance between SNPs:
-  can be number of SNPs in between (i.e. distance based on SNP's raw index in the .bim file)
- can be based on number of base pairs (BP column in .bim file, in Kilobases or Megabases)
-  can be based on genomic position (GP column in .bim file, in centimorgans)
I would allow user to choose his or her favourite alternative same was as LD score regression (https://github.com/bulik/ldsc/blob/master/ldsc.py#L438) i.e. by allowing three options:
-ld-window-snps, -ld-window-kb, -ld-window-cm , with a comment that exactly one of these must be specified (i.e. user can't combine them).  Depending on which option users chooses to specify you'll need to make a slightly different call to snp_cor, passing genetic.dist argument that is based on what user asked for. 

>``If NULL is passed to  infos.pos all column sums of the correlation matrix will be 1 (don't know why).``
One explanation could be that LD matrix is diagonal (each SNPs has 1.0 correlation with itself), and no other correlations were recorded. I.e. the window to computed LD was 0 and no other correlations were considered.

For the warning ``    cat('\nSkipping chromosome', chr,'. Reason: 0 SNPs available\n') ``, I suggest promote this to an error (and skip processing the remaining chromosomes). If users wants to compute LD matrices for subset of chromosomes it's best to use --chr2use argument, otherwise (if the intention is to compute for all chromosomes) it makes little sense to continue if a specific chromosome failed entirely - best for user to investigate what's wrong with input. Also, i think a typical use case will be to run LD computation  in a job array with one task per chromosome. In this case no SNPs 100% indicates a  failure as no output was produced.

I'll select ``request changes`` for the review but all of the above are *nice* to have features for improved usability, though none of them are  *must* have - I'm also fine with merging as it is because calculateLD.R is usable as it is.

**Comment by deepchocolate** [comment]
@espenhgn What are your thoughts on @ofrei s comment regarding `--sumstats` and `--sumstats-sep`? I thought it might relate to your work with the PGS tool.

**Comment by espenhgn** [comment]
> @espenhgn What are your thoughts on @ofrei s comment regarding `--sumstats` and `--sumstats-sep`? I thought it might relate to your work with the PGS tool.

Hi @deepchocolate. My reasons for adding `--sumstats-sep` was not as much related to the PGS tool, it was just a workaround for `utils::read.table` which does not automatically try to assess column separator, and `","` was set as default. Some sumstats files I test with are whitespace separated. Maybe there's another equivalent function that could be used like `data.table::fread`, if `--sumstats` should be kept. 

It's quite possible that `--sumstats` is redundant with the `--file-keep-snps` flag as @ofrei pointed out, but I don't think we shall expect users to know by heart these somewhat cryptic bash calls for extracting the intersect between files. I'm fine with keeping `--sumstats` as an option, as the file is already needed to run the remaining steps in LDpred2 analysis.

**Review by ofrei — APPROVED** [review]

**Comment by ofrei** [comment]
I'm also  fine with keeping --sumstats option. Switching code review to "approved" - as I mentioned all my comments were optional

**Comment by ofrei** [comment]
@deepchocolate thanks for implementing ``--thres-r2``, this one is really good to have!

**Comment by espenhgn** [comment]
@deepchocolate, if you agree, I can remove the `--sumstats-sep` arg I introduced as `dfSumStats <- data.table::fread(fileSumstats, data.table=F)`` appears to give equivalent results as `dfSumStats <- read.table(fileSumstats, header=T, sep=sepSumstats)`.

**Comment by deepchocolate** [comment]
> @deepchocolate, if you agree, I can remove the `--sumstats-sep` arg I introduced as `` dfSumStats <- data.table::fread(fileSumstats, data.table=F)`` appears to give equivalent results as ``dfSumStats <- read.table(fileSumstats, header=T, sep=sepSumstats)`.

Fine with me!

@ofrei I will look at the remaining suggestions in your comment and implement in a separate PR. Including moving the imputation.

**Comment by espenhgn** [comment]
@deepchocolate, you can first merge #129 into this PR? 
Other than that, I think it's all agreed upon for this one?

**Comment by deepchocolate** [comment]
> @deepchocolate, you can first merge #129 into this PR? Other than that, I think it's all agreed upon for this one?

Will do! 
Also, changing `--file-keep-snps` to `--extract` as suggested by @ofrei before merge

**Comment by deepchocolate** [comment]
@espenhgn @ofrei Last thing: Added a note in the README that these scripts currently don't perform any filtering of genotype data like prsice does (MAF/imputation quality) etc.

**Pull request merged**